File: lib/core/problem.dart
/// RFC 7807 Problem Details — the single error currency across the app (mirrors
/// alcohol.argon's `Problem` and the shape zinc returns). House rule: never throw
/// across boundaries — convert to `Result<T>` and carry a `Problem` on failure.
class Problem implements Exception {
  final String type;
  final String title;
  final int status;
  final String? detail;
  final String? instance;

  const Problem({
    required this.type,
    required this.title,
    required this.status,
    this.detail,
    this.instance,
  });

  factory Problem.fromJson(Map<String, dynamic> json) => Problem(
        type: json['type'] as String? ?? 'about:blank',
        title: json['title'] as String? ?? 'Error',
        status: (json['status'] as num?)?.toInt() ?? 0,
        detail: json['detail'] as String?,
        instance: json['instance'] as String?,
      );

  /// Locally-generated problems for failures that never reach the server.
  factory Problem.local(String title,
          {int status = 0, String? detail, String type = 'about:blank'}) =>
      Problem(type: type, title: title, status: status, detail: detail);

  static Problem network(Object error) =>
      Problem.local('Network error', detail: error.toString(), type: 'neon:network');

  static Problem decoding(Object error) =>
      Problem.local('Unexpected response', detail: error.toString(), type: 'neon:decoding');

  static const unauthenticated =
      Problem(type: 'neon:unauthenticated', title: 'Not signed in', status: 401);

  @override
  String toString() => 'Problem($status $title${detail != null ? ': $detail' : ''})';
}

/// Minimal `Result` type — `Ok` on success, `Err` carrying a [Problem] on failure.
/// Use Dart pattern matching: `switch (result) { case Ok(:final value): … }`.
sealed class Result<T> {
  const Result();
}

class Ok<T> extends Result<T> {
  final T value;
  const Ok(this.value);
}

class Err<T> extends Result<T> {
  final Problem problem;
  const Err(this.problem);
}

File: Sources/Core/Problem.swift
import Foundation

/// RFC 7807 Problem Details — the single error currency across the app, mirroring
/// alcohol.argon's `Problem`. The zinc backend returns these on error; we also
/// synthesize local ones (network/decoding/auth) so every failure has the same shape.
///
/// House rule (from argon): never throw across boundaries — convert to `Result<T, Problem>`.
struct Problem: Error, Decodable, Identifiable, Sendable {
  let type: String
  let title: String
  let status: Int
  let detail: String?
  let instance: String?

  var id: String { instance ?? type }

  init(type: String, title: String, status: Int, detail: String? = nil, instance: String? = nil) {
    self.type = type
    self.title = title
    self.status = status
    self.detail = detail
    self.instance = instance
  }

  // Tolerant decoding: zinc Problems are well-formed, but fall back gracefully.
  init(from decoder: Decoder) throws {
    let c = try decoder.container(keyedBy: CodingKeys.self)
    self.type = (try? c.decode(String.self, forKey: .type)) ?? "about:blank"
    self.title = (try? c.decode(String.self, forKey: .title)) ?? "Error"
    self.status = (try? c.decode(Int.self, forKey: .status)) ?? 0
    self.detail = try? c.decodeIfPresent(String.self, forKey: .detail)
    self.instance = try? c.decodeIfPresent(String.self, forKey: .instance)
  }

  private enum CodingKeys: String, CodingKey {
    case type, title, status, detail, instance
  }
}

extension Problem {
  /// Locally-generated problems for failures that never reach the server.
  static func local(_ title: String, status: Int = 0, detail: String? = nil, type: String = "about:blank") -> Problem {
    Problem(type: type, title: title, status: status, detail: detail)
  }

  static func network(_ error: Error) -> Problem {
    .local("Network error", detail: error.localizedDescription, type: "neon:network")
  }

  static func decoding(_ error: Error) -> Problem {
    .local("Unexpected response", detail: error.localizedDescription, type: "neon:decoding")
  }

  static let unauthenticated = Problem.local("Not signed in", status: 401, type: "neon:unauthenticated")
}
